README
Formula Converter Library
=========================


About this version
------------------

This is version 0.9 of the formconv formula converter which is a
successor of the 0.8.x series. What's new in 0.9 is that it has a more
unified documentation and development model --- we made some cleanups
during the development of 0.8.x and finally we released 0.9.0. ;-)

Version 0.8 was a completely rewritten version of 0.7.2 and was not
fully compatible with it. 0.8 and above is a much better implementation
of the formula conversion problem, and is higher configurable than the
older versions. We continously work to add all features from 0.7.2 into
0.8 and above, because some of them currently not yet available (e.g.
"pronouncing" formulas)

If you want to use WebMathematics Interactive 1.x, you may want to use
version 0.7.2.1. In case of using 2.x, version >= 0.8.3 is suggested.

The new version is implemented using ANTLR-2.7.x. (Use version 2.7.7
or better for best results.)

To use version 0.7.2, go to the obsolete/0.7.2/ directory and
read the README file for details.


How to use the library
----------------------

You can use this library to create an application, which may take
transformations on formulae. There are 3 (useful) example
transformations in this package:

PresentationTransform - This is useful if you want to create for
example LaTeX output, or MathML presentation.
			
ContentTransform, ContentTransform2 - These are responsible for creating
the proper output for CAS-es, or a MathML content. These transformations
simplify the structure of the formula and expand the contractions. The
proper order of use is: parser, ContentTransform, ContentTransform2.

To create another formula from the input you have to parse it. (See the
src/cpp for the examples of use.) Currently just IntuitiveParser, 
LispParser, and ContentMathMLInParser are ready to use. After the input
is parsed, you can take transformations on them. After the transformations
you can create an output from it.
(Examples are in src/cpp.)

You can also use the formconv library as a production quality shared
library. A short example is provided in the test/ directory.

Good work!


How to use formconv
-------------------

After building the sources, you will find bin/formconv. This is a
command line interface for the library.

Example 1. You want to convert one of your formula in LaTeX to content
MathML. The following command will create it for you.

-------------------------------------8X-----------------------------------
$ bin/formconv -O mathml --header "<?xml version=\"1.0" \
  encoding="UTF-8" ?>" \
  --prefix "<math xmlns=\"http://www.w3c.org/1998/Math/MathML\">" \
  --suffix "</math>" -o myFirstMathML.xml
  \sum_{j=2}^n{x^j}
-------------------------------------8X-----------------------------------

Note. After the last line please hit EOF: Ctrl-D (on UNIX) or
Ctrl-Z (Windows/DOS).

Example 2. You want to convert your MathML file to a C file. The
following program will make it.

-------------------------------------8X-----------------------------------
$ bin/formconv -O c --header "#include <math.h>
#include <stdio.h>
#define FC_COMPLEX_TYPE myComplex
#define FC_REAL_TYPE double
#define FC_RATIONAL_TYPE myRational
#define FC_INTEGER_TYPE long int
#define FC_VAR_TYPE double
" --prefix "
int main(int arg, char **argv)
{
  printf(\"%g\\n\"," --suffix ")
}" -o myFirstCProgram.c -i myFirstMathML.xml
-------------------------------------8X-----------------------------------

Note. The above is one command. If you want to compile, try it with 

$ gcc -o myFirstCProgram myFirstCProgram.c

To run it:

$ ./myFirstCProgram

The abilities of the program can be asked with:

$ bin/formconv --help

We hope you like this interface.


HISTORY
Version 0.9.2 (yet unreleased)

  Fixing a bug when linking the library

Version 0.9.1 (2010-02-22)

  Bugfixes for WMI2-2.1.6 (see ChangeLog).

Version 0.9.0 (2008-12-15)

  Regression suite including MAW and WMI2 inputs.
  Bugfixes for WMI2-2.1.5 (see ChangeLog).
  Updated manual page.
  Rewritten OVERVIEW (redrawn).
  Library numbering.

Version 0.8.8 (2008-07-12)

  Support for rationals and some bug fixes (e^x^2).
  libboost-dev is needed for compilation.

Version 0.8.7 (2008-03-18)

  Support for tclap-1.1.0.

Version 0.8.6 (2008-03-15)

  Bugfixes for WMI-2.1.2 (see ChangeLog).

Version 0.8.5 (2008-02-22)

  Mostly bugfixes for WMI-2.1.1 and Robert Marik's math projects.
  Detection of runantlr (formerly antlr) executable.

Version 0.8.4 (January 2008)

  Second public release for WMI-2.1.1.

Version 0.8.3 (December 2007)

  First public release especially for WebMathematics Interactive 2.1.

Version 0.8.2 (2005-08-13)

  Introducing constants FC_TRUE and FC_FALSE for compatibility.

Version 0.8 (2004-11-23)

  First public release of libformconv.


src/cpp/formconv.cpp
/*
 * General formula converter.
 * Command line option parser.
 * A demonstration what the library can do.
 * $Id: formconv.cpp,v 1.36 2010/08/28 13:20:24 kovzol Exp $
 */

#include <iostream>
#include <fstream>
#include <tclap/CmdLine.h>
#include <map>

#include <h/misc.h>
#include <h/fcAST.h>
#include <h/fcversion.h>

#include <COutput.hpp>
#include <CComplexOutput.hpp>
#include <ComplexCompute.hpp>
#include <ContMathMLInLexer.hpp>
#include <ContMathMLInParser.hpp>
#include <ContMathMLToIntuitiveTransform.hpp>
#include <ContentTransform.hpp>
#include <ContentTransform2.hpp>
#include <ConstantTransform.hpp>
#include <IntuitiveLexer.hpp>
#include <IntuitiveParser.hpp>
#include <JavaOutput.hpp>
#include <LaTeXOutput.hpp>
#include <LispLexer.hpp>
#include <LispParser.hpp>
#include <LispOutput.hpp>
#include <MathMLOutput.hpp>
#include <MathematicaOutput.hpp>
#include <MaximaOutput.hpp>
#include <MapleOutput.hpp>
#include <GnuplotOutput.hpp>
#include <MupadOutput.hpp>
#include <HTMLOutput.hpp>
#include <PresentationTransform.hpp>
#include <Transform.hpp>

#ifdef TCLAP_CONSTRAINT_H
#define ARGS(BOOL, CMD) CMD, BOOL
#else
#define ARGS(BOOL, CMD) BOOL, CMD
#endif

using namespace TCLAP;
using namespace std;
using namespace antlr;

int main (int argc, char **argv)
{
	string inputFileName, outputFileName, inputFormat, outputFormat, prefix,
		suffix, header, footer, inputContent, tmp;
	ThreeState times;
	bool dynamicParentheses, beautyParentheses, autoSimplify, cPrecedence, typeCheck, disallowComputingFunctions, noLongVars, forceFloatDivision;
	vector<string> functionNames, changes, variable, allowedLetters;
	Language inputLanguage, outputLanguage;

	try {
		//Define command line object
		CmdLine cmd("Formula Converter (formconv)\nCopyright (C) 2003-2010 Zoltan Kovacs <kovzol@math.u-szeged.hu>\n"
		    "Copyright (C) 2003-2010 Gabor Bakos <baga@users.sourceforge.net>\n"
		    "This is free software with ABSOLUTELY NO WARRANTY.", ' ', FCVERSION_STRING);
		ValueArg<string> inputFileNameArg("i", "inputfile", "Specifies the input filename. (Default: standard input)", false,
#ifdef WIN32
		"con"
#else
		"/dev/stdin"
#endif
		, "string", cmd);
		ValueArg<string> outputFileNameArg("o", "outputfile", "Specifies the output filename. (Default: standard output)", false,
#ifdef WIN32
		"con"
#else
		"/dev/stdout"
#endif
		, "string", cmd);
		vector<string> allowedInputFormat;
		allowedInputFormat.push_back("intuitive");
		allowedInputFormat.push_back("contentMathML");
		allowedInputFormat.push_back("lisp");
#ifdef TCLAP_CONSTRAINT_H
		ValuesConstraint<string> constraintInputFormat(allowedInputFormat);
		ValueArg<string> inputFormatArg("I", "inputformat", "Specifies the format of the input.", false, "intuitive", &constraintInputFormat, cmd);
#else
		ValueArg<string> inputFormatArg("I", "inputformat", "Specifies the format of the input.", false, "intuitive", allowedInputFormat, cmd);
#endif
		vector<string> allowedOutputFormat;
		allowedOutputFormat.push_back("c");
		allowedOutputFormat.push_back("ccomplex");
		allowedOutputFormat.push_back("complexcompute");
		allowedOutputFormat.push_back("java");
		allowedOutputFormat.push_back("latex");
		allowedOutputFormat.push_back("mathematica");
		allowedOutputFormat.push_back("mathml");
		allowedOutputFormat.push_back("lisp");
		allowedOutputFormat.push_back("maxima");
		allowedOutputFormat.push_back("maple");
		allowedOutputFormat.push_back("gnuplot");
		allowedOutputFormat.push_back("mupad");
		allowedOutputFormat.push_back("html");
		allowedOutputFormat.push_back("debug");
		allowedOutputFormat.push_back("debugContent");
#ifdef TCLAP_CONSTRAINT_H
		ValuesConstraint<string> constraintOutputFormat(allowedOutputFormat);
		ValueArg<string> outputFormatArg("O", "ouputformat", "Specifies the format of the output.", false, "latex", &constraintOutputFormat, cmd);
#else
		ValueArg<string> outputFormatArg("O", "ouputformat", "Specifies the format of the output.", false, "latex", allowedOutputFormat, cmd);
#endif
		SwitchArg dynamicParenthesesArg("d", "dynamicparentheses", "Uses in LaTeX output \\left(, and \\right), ...", ARGS(true, cmd));
		vector<string> threeStateType;
		threeStateType.push_back("space");
		threeStateType.push_back("automatic");
		threeStateType.push_back("cdot");
#ifdef TCLAP_CONSTRAINT_H
		ValuesConstraint<string> constraintThreeStateType(threeStateType);
		ValueArg<string> timesArg("t", "times", "Prints space in presentation outputs instead of *, or \\cdot, depending on the value.", false, "automatic", &constraintThreeStateType, cmd);
#else
		ValueArg<string> timesArg("t", "times", "Prints space in presentation outputs instead of *, or \\cdot, depending on the value.", false, "automatic", threeStateType, cmd);
#endif
		SwitchArg beautyParenthesesArg("b", "beautyparentheses", "Uses in presentation outputs (, ), [, ], {, } instead of just (, ).", ARGS(false, cmd));
		SwitchArg autoSimplifyArg("a", "autosimplify", "Does not remove the unnecessary parentheses from the formulae.", ARGS(true, cmd));
//		SwitchArg pascalPrecedenceArg("P", "Pascalprecedence", "Uses Pascal precedence in input (else C precedence).", true, cmd);
//		SwitchArg typeCheckArg("t", "type-check", "Checks, whether the arguments can be used to the given operator.", false, cmd);
		SwitchArg disallowComputingFunctionsArg("D", "disallowcomputingfunctions", "Forbids the use of functions like sum, prod, int, diff...", ARGS(false, cmd));
		SwitchArg noLongVarsArg("n", "nolongvars", "Forbids the use of variable/parameter names longer than 1 character.", ARGS(false, cmd));
		MultiArg<string> functionNamesArg("f", "functionnames", "Use these names as user defined function names. (In input you are/will able to define functions with other names.)", false, "string", cmd);
//		ValueArg<string> inputContentArg("C", "inputcontent", "Specifies a restriction on the format of intuitive input.", false, "set-expression", allowedContentType, cmd);
		MultiArg<string> allowedLettersArg("A", "allowedletters", "Just these letters of accepted as variables or parameters in the input.", false, "string", cmd);
		ValueArg<string> headerArg("H", "header", "Prints this text before the output.", false, "", "string", cmd);
		ValueArg<string> footerArg("F", "footer", "Prints this text after the output.", false, "", "string", cmd);
		ValueArg<string> prefixArg("p", "prefix", "Prints this text before each formula.", false, "", "string", cmd);
		ValueArg<string> suffixArg("s", "suffix", "Prints this text after each formula.", false, "", "string", cmd);
		SwitchArg forceFloatDivArg("r", "forceFloatDiv", "Always selects the floating point division for the output of applicable.", ARGS(false, cmd));

		vector<string> languageType;
		languageType.push_back("hungarian");
		languageType.push_back("english");
		languageType.push_back("german");
		languageType.push_back("french");
		languageType.push_back("simple");
#ifdef TCLAP_CONSTRAINT_H
		ValuesConstraint<string> constraintLanguageType(languageType);
		ValueArg<string> inputLanguageArg("l", "inputlanguage", "Uses this language's specialties in the input.", false, "english", &constraintLanguageType, cmd);
		ValueArg<string> outputLanguageArg("L", "outputlanguage", "Uses this language's specialties in the output.", false, "english", &constraintLanguageType, cmd);
#else
		ValueArg<string> inputLanguageArg("l", "inputlanguage", "Uses this language's specialties in the input.", false, "english", languageType, cmd);
		ValueArg<string> outputLanguageArg("L", "outputlanguage", "Uses this language's specialties in the output.", false, "english", languageType, cmd);
#endif
		MultiArg<string> changesArg("x", "changename", "Changes the names of the variables/parameters to the given. (Ex.: -x x,x4 will change all x variable to x4) (Please use just one comma per rule.)", false, "string", cmd);
		MultiArg<string> variableArg("V", "variable", "Sets a variable in a computing output. (Ex.: -V x=4) (Please do not use spaces in the expression.)", false, "string", cmd);

		cmd.parse(argc, argv);
		inputFileName = inputFileNameArg.getValue();
		outputFileName = outputFileNameArg.getValue();
		inputLanguage = formconvLanguage::getLanguage(inputLanguageArg.getValue());
		outputLanguage = formconvLanguage::getLanguage(outputLanguageArg.getValue());
		inputFormat = inputFormatArg.getValue();
		outputFormat = outputFormatArg.getValue();
		dynamicParentheses = dynamicParenthesesArg.getValue();
		tmp = timesArg.getValue();
		if (tmp == "space")
			times = no;
		else if (tmp == "automatic")
			times = automatic;
		else if (tmp == "cdots")
			times = yes;
		else
			throw std::exception(); //This should not happen
		beautyParentheses = beautyParenthesesArg.getValue();
		autoSimplify = autoSimplifyArg.getValue();
//		cPrecedence = !(pascalPrecedenceArg.getValue());
//		typeCheck = typeCheckArg.getValue();
		disallowComputingFunctions = disallowComputingFunctionsArg.getValue();
		noLongVars = noLongVarsArg.getValue();
		forceFloatDivision = forceFloatDivArg.getValue();
		functionNames = functionNamesArg.getValue();
//		inputContent = inputContentArg.getValue();
		allowedLetters = allowedLettersArg.getValue();
		header = headerArg.getValue();
		footer = footerArg.getValue();
		prefix = prefixArg.getValue();
		suffix = suffixArg.getValue();
		changes = changesArg.getValue();
		variable = variableArg.getValue();
	} catch (ArgException &e)
	{
		cerr << "Error: " << e.error() << " in argument: " << e.argId() << endl;
	}
	try {
		//Setting the proper input and output streams
		ifstream inputStream(inputFileName.c_str());
		ofstream outputStream(outputFileName.c_str());
		//Creating the AST factory
		ASTFactory myFactory("FcAST", FcAST::factory);
		RefFcAST ast;
		if (inputFormat == "intuitive")
		{
			//Create the lexer
			IntuitiveLexer lexer(inputStream);
			lexer.setLanguage(inputLanguage);
			lexer.setLongVars(!noLongVars);
			
			//Setting the acceptable variable to the lexer
			for (vector<string>::iterator it = allowedLetters.begin(); it != allowedLetters.end(); ++it)
			{
				lexer.addAllowedLetter(*it);
			}
			//Create the parser
			IntuitiveParser parser(lexer);
			parser.initializeASTFactory(myFactory);
			parser.setASTFactory(&myFactory);
			//Setting the parameters for the parser
			for (vector<string>::iterator it = functionNames.begin(); it != functionNames.end(); ++it)
			{
				parser.addFunctionName(*it);
			}
			//Sets wheter you want to use sum, prod, int, diff, ...
			parser.setAllowComputingFunctions(!disallowComputingFunctions);
			//Start parsing
			parser.expr();
			//Get AST for further processing
			ast = RefFcAST(parser.getAST());
			if (outputFormat == "debug" || outputFormat == "debugContent")
			{
				if (ast)
					outputStream << "\nIntuitive input, before transformations.\n" << ast->toStringList();
				else
					outputStream << "null AST\n";
				outputStream.flush();
			}
		}
		else if (inputFormat == "contentMathML")
		{
			//Create the lexer
			ContMathMLInLexer lexer(inputStream);
			//Create the parser
			ContMathMLInParser parser(lexer);
			parser.initializeASTFactory(myFactory);
			parser.setASTFactory(&myFactory);
			//Start parsing
			parser.inp();
			//Get AST for further processing
			ast = RefFcAST(parser.getAST());
			if (outputFormat == "debug" || outputFormat == "debugContent")
			{
				if (ast)
					outputStream << "\nMathML input, before transformations.\n" << ast->toStringList();
				else
					outputStream << "null AST\n";
				outputStream.flush();
			}
			//Convert it to Intuitive AST format to able to process with other transforms, outputs.
			ContMathMLToIntuitiveTransform m2i;
			m2i.setASTFactory(&myFactory);
			m2i.inp(ast);
			ast = RefFcAST(m2i.getAST());
			if (outputFormat == "debug" || outputFormat == "debugContent")
			{
				if (ast)
					outputStream << "\nMathML input, before transformations.\n" << ast->toStringList();
				else
					outputStream << "null AST\n";
				outputStream.flush();
			}
		}
		else if (inputFormat == "lisp")
		{
			//Create the lexer
			LispLexer lexer(inputStream);
			lexer.init();
			//Create the parser
			LispParser parser(lexer);
			parser.initializeASTFactory(myFactory);
			parser.setASTFactory(&myFactory);
			//Start parsing
			parser.inp();
			//Get AST for further processing
			ast = RefFcAST(parser.getAST());
			//Convert it to Intuitive AST format to able to process with other transforms, outputs.
//			ContMathMLToIntuitiveTransform m2i;
//			m2i.setASTFactory(&myFactory);
//			m2i.inp(ast);
//			ast = RefFcAST(m2i.getAST());
			if (outputFormat == "debug" || outputFormat == "debugContent")
			{
				if (ast)
					outputStream << "\nMathML input, before transformations.\n" << ast->toStringList();
				else
					outputStream << "null AST\n";
				outputStream.flush();
			}
		}
		else
		{
			throw std::exception(); //This should not happen
		}
		//We have the result in AST, let's transform it to the proper form
		if (outputFormat == "latex" || outputFormat == "html" || outputFormat == "debug")
		{
			//Create the presentation transformator
			PresentationTransform pt;
			//Set the properties
			pt.setSimplify(autoSimplify);
			pt.setSimplifyPower(true);
			pt.setInvisibleTimes(times);
			pt.setTokenPrecedence(PresentationTransform::presentationTokenPrecedence);
			if (outputFormat == "latex")
			{
				pt.setDivAsFrac(yes);
			}
			else if (outputFormat == "html")
			{
				pt.setDivAsFrac(automatic);
			}
			else
			{
				pt.setDivAsFrac(automatic);
			}
			pt.initializeASTFactory(myFactory);
			pt.setASTFactory(&myFactory);
			//Do the transformation
			pt.inp(ast);
			ast = RefFcAST(pt.getAST());
			if (outputFormat == "debug")
			{
				if (ast)
					outputStream << "\nAfter presentation transformation.\n" << ast->toStringList() << "\n";
				else
					outputStream << "null AST\n";
				outputStream.flush();
			}
		} else //Not presentation transform
		{
			//Create the first transformation
			ContentTransform ct;
			ct.initializeASTFactory(myFactory);
			ct.setASTFactory(&myFactory);
			//Apply the transformation
			ct.inp(ast);
			//Get the result
			ast = RefFcAST(ct.getAST());
			if (outputFormat == "debugContent")
			{
				if (ast)
					outputStream << "\nAfter first content transformation.\n" << ast->toStringList();
				else
					outputStream << "null AST\n";
				outputStream.flush();
			}
			//Create the second transformation
			ContentTransform2 ct2;
			ct2.setASTFactory(&myFactory);
			//Apply the transformation
			ct2.inp(ast);
			//Get the result
			ast = RefFcAST(ct2.getAST());
			if (outputFormat == "debugContent")
			{
				if (ast)
					outputStream << "\nAfter second content transformation.\n" << ast->toStringList() << "\n";
				else
					outputStream << "null AST\n";
				outputStream.flush();
			}
		}
		if (!changes.empty())
		{
			//Create the transformator for (currently just) for changes
			Transform t;
			t.initializeASTFactory(myFactory);
			t.setASTFactory(&myFactory);
			for (vector<string>::iterator it = changes.begin(); it != changes.end(); ++it)
			{
				string::size_type pos = (*it).find(',');
				t.addChange((*it).substr(0, pos), (*it).substr(pos+1));
			}
			t.inp(ast);
			ast = RefFcAST(t.getAST());
			if (outputFormat == "debug" || outputFormat == "debugContent")
			{
				if (ast)
					outputStream << "\nAfter variable name changing transformation.\n" << ast->toStringList() << "\n";
				else
					outputStream << "null AST\n";
				outputStream.flush();
			}
		}
		//ast now contains the proper AST
		//Now creating the output
		outputStream << header;
		if (outputFormat == "latex")
		{
			//Create the proper object
			LaTeXOutput latex;
			latex.initializeASTFactory(myFactory);
			latex.setASTFactory(&myFactory);
			//Set the properties
			latex.setLanguage(outputLanguage);
			latex.setDynamicParentheses(dynamicParentheses);
			latex.setBeautyParentheses(beautyParentheses);
			//Get the result
			string s = latex.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
		}
		else if (outputFormat == "html")
		{
			//Create the proper object
			HTMLOutput html;
			html.initializeASTFactory(myFactory);
			html.setASTFactory(&myFactory);
			//Set the properties
			html.setLanguage(outputLanguage);
			html.setDynamicParentheses(dynamicParentheses);
			html.setBeautyParentheses(beautyParentheses);
			//Get the result
			string s = html.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		else if (outputFormat == "c")
		{
			//Create the proper object
			COutput c;
			c.initializeASTFactory(myFactory);
			c.setASTFactory(&myFactory);
			c.setForceFloatDivision(forceFloatDivision);
			c.init();
			//Get the result
			string *s = c.inp(ast);
			string *decl = c.getDeclarations();
			string *func = c.getFunctions();

			outputStream << *decl << *func;
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << *s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			
			//delete the unnecessary variables...
			delete s;
			delete decl;
			delete func;
		}
		else if (outputFormat == "ccomplex")
		{
			//Create the proper object
			CComplexOutput c;
			c.initializeASTFactory(myFactory);
			c.setASTFactory(&myFactory);
			c.setForceFloatDivision(forceFloatDivision);
			c.init();
			//Get the result
			string *s = c.inp(ast);
			string *decl = c.getDeclarations();
			string *func = c.getFunctions();

			outputStream << *decl << *func;
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << *s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			
			//delete the unnecessary variables...
			delete s;
			delete decl;
			delete func;
		}
		else if (outputFormat == "java")
		{
			//Create the proper object
			JavaOutput java;
			java.initializeASTFactory(myFactory);
			java.setASTFactory(&myFactory);
			//Get the result
			string s = java.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		else if (outputFormat == "mathematica")
		{
			//Create the proper object
			MathematicaOutput mathematica;
			mathematica.initializeASTFactory(myFactory);
			mathematica.setASTFactory(&myFactory);
			//Get the result
			string s = mathematica.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		else if (outputFormat == "mathml")
		{
			//Create the proper object
			MathMLOutput mathml;
			mathml.initializeASTFactory(myFactory);
			mathml.setASTFactory(&myFactory);
			mathml.init();
			//Get the result
			string s = mathml.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		else if (outputFormat == "lisp")
		{
			//Create the proper object
			LispOutput lisp;
			lisp.initializeASTFactory(myFactory);
			lisp.setASTFactory(&myFactory);
			//Get the result
			string s = lisp.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		else if (outputFormat == "complexcompute")
		{
			//Create the proper object
			ComplexCompute cc;
			cc.initializeASTFactory(myFactory);
			cc.setASTFactory(&myFactory);
			//Set the given variables
			for (vector<string>::iterator it = variable.begin(); it != variable.end(); ++it)
			{
				std::istringstream varinput((*it).substr((*it).find('=')+1));
				IntuitiveLexer varLexer(varinput);
				varLexer.setLanguage(inputLanguage);
				IntuitiveParser varParser(varLexer);
				varParser.initializeASTFactory(myFactory);
				varParser.setASTFactory(&myFactory);
				varParser.expr();
				complex<double> value = cc.inp(RefFcAST(varParser.getAST()));
				cc.setValue((*it).substr(0, (*it).find('=')), value);
			}
			//Get the result
			complex<double> z = cc.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << z.real();
			if (abs(z.imag())>1E-40)
				outputStream << "+" << z.imag() << "i";
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		else if (outputFormat == "maxima")
		{
			//Create the proper object
			MaximaOutput maxima;
			maxima.initializeASTFactory(myFactory);
			maxima.setASTFactory(&myFactory);
			//Get the result
			string s = maxima.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		else if (outputFormat == "maple")
		{
			//Create the proper object
			MapleOutput maple;
			maple.initializeASTFactory(myFactory);
			maple.setASTFactory(&myFactory);
			//Get the result
			string s = maple.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		else if (outputFormat == "gnuplot")
		{
			//Create the constant transformation
			ConstantTransform constTransform;
			constTransform.setASTFactory(&myFactory);
			constTransform.setConvertDecimal(true);
			constTransform.setConvertExponents(true);
			//Apply the transformation
			constTransform.inp(ast);
			//Get the result
			ast = RefFcAST(constTransform.getAST());
			if (outputFormat == "debugContent")
			{
				if (ast)
					outputStream << "\nAfter second constant transformation.\n" << ast->toStringList() << "\n";
				else
					outputStream << "null AST\n";
				outputStream.flush();
			}
			//Create the proper object
			GnuplotOutput gnuplot;
			gnuplot.initializeASTFactory(myFactory);
			gnuplot.setASTFactory(&myFactory);
			gnuplot.setForceFloatDivision(forceFloatDivision);
			//Get the result
			string s = gnuplot.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		else if (outputFormat == "mupad")
		{
			//Create the proper object
			MupadOutput mupad;
			mupad.initializeASTFactory(myFactory);
			mupad.setWriteParen(false);
			mupad.setASTFactory(&myFactory);
//			outputStream << "\nAfter variable name changing transformation.\n" << ast->toStringList();
			if (autoSimplify)
			{
		        	//Create the presentation transformator
				PresentationTransform pt;
				//Set the properties
				pt.setSimplify(yes);
				pt.setSimplifyPower(false);
				pt.setInvisibleTimes(no);
				pt.setDivAsFrac(no);
				pt.setTokenPrecedence(MupadOutput::mupadTokenPrecedence);
				pt.initializeASTFactory(myFactory);
				pt.setASTFactory(&myFactory);
				//Do the transformation
				pt.inp(ast);
				ast = RefFcAST(pt.getAST());
//				outputStream << "\nAfter variable name changing transformation.\n" << ast->toStringList();
			}
			//Get the result
			string s = mupad.inp(ast);
			//TODO change it to the outputs
			outputStream << prefix;
			outputStream << s;
			//TODO change it to the outputs
			outputStream << suffix << endl;
			outputStream.flush();
		}
		outputStream << footer;
	} catch (ANTLRException& e)
	{
		cerr << "Exception: " << e.toString() << endl;
		return -1;
	} catch (exception& e)
	{
		cerr << "Exception: " << e.what() << endl;
		return -1;
	}
}
